clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -90,24 +89,279 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+      if (opts.emitImplicitIntegerSignChangeChecks)
+        cgf.getCIRGenModule().errorNYI(loc,
+                                       "implicit integer sign change checks");
       return src;
+    }
 
-    cgf.getCIRGenModule().errorNYI(loc,
-                                   "emitScalarConversion for unequal types");
-    return {};
+    if (dstType->isVoidType())
+      return nullptr;
+
+    mlir::Type mlirSrcType = src.getType();
+
+    // Handle conversions to bool first, they are special: comparisons against
+    // 0.
+    if (dstType->isBooleanType())
+      return emitConversionToBool(src, srcType, cgf.getLoc(loc));
+
+    mlir::Type mlirDstType = cgf.convertType(dstType);
+
+    if (srcType->isHalfType() &&
+        !cgf.getContext().getLangOpts().NativeHalfType) {
+      // Cast to FP using the intrinsic if the half type itself isn't supported.
+      if (mlir::isa<cir::CIRFPTypeInterface>(mlirDstType)) {
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics())
+          cgf.getCIRGenModule().errorNYI(loc,
+                                         "cast via llvm.convert.from.fp16");
+      } else {
+        // Cast to other types through float, using either the intrinsic or
+        // FPExt, depending on whether the half type itself is supported (as
+        // opposed to operations on half, available with NativeHalfType).
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics()) {
+          cgf.getCIRGenModule().errorNYI(loc,
+                                         "cast via llvm.convert.from.fp16");
+          // FIXME(cir): For now lets pretend we shouldn't use the conversion
+          // intrinsics and insert a cast here unconditionally.
+        }
+        src = builder.createCast(cgf.getLoc(loc), cir::CastKind::floating, src,
+                                 cgf.FloatTy);
+        srcType = cgf.getContext().FloatTy;
+        mlirSrcType = cgf.FloatTy;
+      }
+    }
+
+    // TODO(cir): LLVM codegen ignore conversions like int -> uint,
+    // is there anything to be done for CIR here?
+    if (mlirSrcType == mlirDstType) {
+      if (opts.emitImplicitIntegerSignChangeChecks)
+        cgf.getCIRGenModule().errorNYI(loc,
+                                       "implicit integer sign change checks");
+      return src;
+    }
+
+    // Handle pointer conversions next: pointers can only be converted to/from
+    // other pointers and integers. Check for pointer types in terms of LLVM, as
+    // some native types (like Obj-C id) may map to a pointer type.
+    if (isa_cast<cir::PointerType>(mlirDstType)) {
+      cgf.getCIRGenModule().errorNYI(loc, "pointer casts");
+      CharUnits alignment = cgf.getContext().getTypeAlignInChars(dstType);
+      auto addr =
+          builder.createAlloca(src.getLoc(), builder.getPointerTo(dstPT), dstPT,
+                               {}, cgf.cgm.getSize(alignment));
+      return builder.createLoad(src.getLoc(), addr);
```

Comment:
Wonder if there is use of a `createDummyValue` function in builder that does all this for us?